### PR TITLE
HZDR gitlab CI

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Use n-wise testing to select test configurations for the compiler, cuda,
boost version.

- add n-wise matrix generator `n_wise_generator.py`
  - add filter for invalid combinations e.g. compiler, boost, cuda
- build dynamic jobs based on the reduced matrix

### CODE REVIEW

File: share/ci/n_wise_generator.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+for i in clang_compiers:
+    cuda_nvcc_compilers.append(i + ("nvcc",))
+for i in gnu_compilers:
+    cuda_nvcc_compilers.append(i + ("nvcc",))
+compilers.append(cuda_nvcc_compilers)
+
+# PIConGPU backend list
+# tuple with two components (backend name, version)
+# version is only required for the cuda backend
+backends = [  ("cuda", 9.2), ("cuda", 10.0), ("cuda", 10.1), ("cuda", 10.2), ("omp2b",), ("single",)]
```
Comment (sbastrakov) [inline share/ci/n_wise_generator.py:132]: Discussed offline: should be `serial`

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -1,100 +1,45 @@
...
+  script:
+    - source $CI_PROJECT_DIR/share/ci/git_rebase.sh
 
-# clang matrix
-matrix-clang:
+# generate reduced test matrix
+# required variables (space separated lists):
+#   PIC_INPUTS - path to examples relative to share/picongpu
+#                e.g.
+#                    "examples" starts one gitlab job per director in `examples/*`
```
Comment (sbastrakov) [inline .gitlab-ci.yml:30]: Should be `directory`

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -1,100 +1,45 @@
...
-    PIC_INPUTS: "examples/Empty"
-    GITLAB_BASES: ".nvcc_cuda102"
-    CXX_VERSIONS: "g++"
-    BOOST_VERSIONS: "1.65.1"
-    PIC_ACCS: "cuda:35"
-  extends: ".base_generator" 
-    
-compile-matrix-nvcc-cuda102:
-  stage: build
+compile-matrix-pair:
```
Comment (sbastrakov) [inline .gitlab-ci.yml:39]: Perhaps "combination" is a better name than "pair", as we are basically talking about one row in the test matrix (after everything is enumerated)? Here and in other places.

File: share/ci/generate_reduced_matrix.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+set -e
+set -o pipefail
+
+# generate a job reduced matrix based on the environment variable lists (space separated)
```
Comment (sbastrakov) [inline share/ci/generate_reduced_matrix.sh:6]: `generate a job reduced` - probably, some preposition or another word is missing after `job`

File: share/ci/git_rebase.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+echo "curl -w -u psychocoderHPC:$GITHUB_TOKEN -X GET https://api.github.com/repos/${github_group_repo}/pulls/${pr_id} 2>/dev/null"
+curl_data=$(curl -u psychocoderHPC:$GITHUB_TOKEN -X GET https://api.github.com/repos/${github_group_repo}/pulls/${pr_id} 2>/dev/null)
+echo "curl data"
+echo "$curl_data"
+# get the destination branch
+target_branch=$(echo "$curl_data" | python3 -c 'import json,sys;obj=json.loads(sys.stdin.read());print(obj["base"]["ref"])')
+echo "target_branch=${target_branch}"
+
+mainline_exists=$(git remote -v | cut -f1 | grep mainline -q && echo 0 || echo 1)
+# avoid that we add the remote reposetory twice if gitlab already cached this operation
```
Comment (sbastrakov) [inline share/ci/git_rebase.sh:23]: Typos, can be re-written as `# avoid adding the remote repository twice if gitlab already cached this operation`

**Comment by sbastrakov** [comment]
@psychocoderHPC is it ready for re-review? Please ping me when it is.

**Comment by psychocoderHPC** [comment]
The failing clang10 native CUDA compile test will be fixed with  #3310

**Comment by sbastrakov** [comment]
@psychocoderHPC thanks, I've just merged #3310 .

**Comment by psychocoderHPC** [comment]
I pushed again some changes:
- introduced a environment variable `CI_MAX_PARALLELISM` we can set in the ci web interface to avoid that the nodes running out of memory. I reduced the parallelism from  8 to 7 now because from time to time some jobs crashed.

**Pull request merged**